mvlearn/utils/utils.py

```diff
@@ -155,3 +172,345 @@ def check_Xs_y_nan_allowed(
+            View idx data.
+        idx : int
+            0 if view 1. 1 if view 2.
+
+        Returns
+        -------
+        result : nd-array
+            Result of linear transformation on input data.
+        """
+        result = x - self.m[idx].reshape([1, -1]).repeat(len(x), axis=0)
+        result = np.dot(result, self.w[idx])
+        return result
+
+    def test(self, H1, H2):
+        """
+        Transform inputs based on already fit matrices.
+
+        Parameters
+        ----------
+        H1 : nd-array, shape (n_samples, n_features)
+            View 1 data.
+        H2 : nd-array, shape (n_samples, n_features)
+            View 2 data.
+
+        Returns
+        -------
+        Results : list, length=2
+            Results of linear transformation on input data.
+        """
+        return [self._get_result(H1, 0), self._get_result(H2, 1)]
+
+
+class cca_loss():
+    """
+    An implementation of the loss function of linear CCA as introduced
+    in the original paper for ``DCCA`` [#1Utils]_.
+
+    Parameters
+    ----------
+    outdim_size : int (positive)
+        The output dimensionality of the CCA transformation.
+    use_all_singular_values : boolean
+        Whether or not to use all the singular values in the loss calculation.
+        If False, only use the top outdim_size singular values.
+    device : torch.device object
+        The torch device being used in DCCA.
+
+    Attributes
+    ----------
+    outdim_size : int (positive)
+        The output dimensionality of the CCA transformation.
+    use_all_singular_values : boolean
+        Whether or not to use all the singular values in the loss calculation.
+        If False, only use the top outdim_size singular values.
+    device : torch.device object
+        The torch device being used in DCCA.
+
+    References
+    ----------
+    .. [#1Utils] Andrew, G., Arora, R., Bilmes, J., & Livescu, K. (2013,
+                 February). Deep canonical correlation analysis. In
+                 International conference on machine learning (pp. 1247-1255).
+    """
+    def __init__(self, outdim_size, use_all_singular_values, device):
+        self.outdim_size = outdim_size
+        self.use_all_singular_values = use_all_singular_values
+        self.device = device
+
+    def loss(self, H1, H2):
+        """
+        Compute the loss (negative correlation) between 2 views.
+
+        Parameters
+        ----------
+        H1: torch.tensor, shape (n_samples, n_features)
+            View 1 data.
+        H2: torch.tensor, shape (n_samples, n_features)
+            View 2 data.
+        """
+
+        r1 = 1e-3
+        r2 = 1e-3
+        eps = 1e-9
+
+        H1, H2 = H1.t(), H2.t()
+
+        o1 = o2 = H1.size(0)
+
+        m = H1.size(1)
+
+        H1bar = H1 - H1.mean(dim=1).unsqueeze(dim=1)
+        H2bar = H2 - H2.mean(dim=1).unsqueeze(dim=1)
+
+        SigmaHat12 = (1.0 / (m - 1)) * torch.matmul(H1bar, H2bar.t())
```

Comment:
A lot of math going on here. Can you provide more comments on what the loss function is either here or in the class documentation?

Reply:
Added extensive documentation in the Notes section of DCCA which will render online, and added a few line-by-line comments briefly explaining code.

Reply:
Documentation of DCCA includes extensive explanation, and small comments in line